### ISSUE TITLE
Private Unit Tests Rule: Check Return Value of Test?

### ISSUE THREAD

**Issue by cfilipov** [realm/SwiftLint#782]
Related to #761.

This is more of a question for the community and maintainers than an actual bug, but you be the judge:

I found a "false positive" using this rule. In the code base I'm working on I observed something like this:

``` Swift
class MyTests: XCTest {
    func testFoo() { ... } // This is an actual test

    private func testableBar() -> Bar {...} // This is just a helper method
}
```

The private unit test rule will treat the `testableBar()` function as a violation because it's a method on an `XCTestCase` subclass which starts with "test" and is marked `private`. I haven't checked to see if XCTest treats methods whose name begins with "test" but return a value as a test. IIRC tests must begin with the word "test" **and** return `Void`, this is certainly the case for Linux XCTests (looking at the type for `allTests`).

I could update the logic to check for `Void`, however... maybe this is actually something that **should** be a style violation? Consider the following: The code I ran into was actually worse than the example above. `testableBar()` was actually named `testBar()`. Again, this wasn't a unit test, it was a helper method for returning a mocked version of `Bar` for testing. This is way too close to what a test looks like, and I had to read the code carefully to determine if perhaps this actually was meant to be a test but someone returned a value by mistake, not realizing this makes it no longer a test.

I think it would be too complex and error-prone to try and differentiate between `testableBar()` and `testBar()`. I think the current behavior is actually desirable, and that both examples are bad practice. If something looks too similar to a unit test, but isn't actually a unit test, that should be a style violation (any method that starts with "test" but doesn't return `Void`). The same reasoning behind having this rule in the first place. You have a method that looks like it tests some code, but if you mistakenly mark it `private`, or make it return something, then it will no longer get run.

Thoughts?

**Comment by jpsim** [comment]
interesting... actually I think I have code that follows this approach, though I don't know if I called the helpers `func test...` or something else. IMHO that's a smell, since it could easily be confused as a "real" test, and then you'd be sad if it doesn't actually run.

**Comment by cfilipov** [comment]
I'd suggest this be marked as "question" rather than "enhancement" since it's debatable whether this is a change that should be made. It's technically more correct to check for "Void" return but the lack of check seems better as a rule. Feel free to close this.

**Comment by jpsim** [comment]
I added the `discussion` tag in case any one wants to look into this further and share their findings.
